src/main.rs

```diff
@@ -1,60 +1,254 @@
+
+    let mut timestamps = Vec::new();
+    for i in block_height - 10..block_height {
+        let block = get_block_by_height(i)?;
+        timestamps.push(block.header.time);
+    }
+
+    let mut total_diff = 0;
+    for i in 1..timestamps.len() {
+        total_diff += timestamps[i] - timestamps[i - 1];
+    }
+
+    let avg_diff = total_diff / (timestamps.len() as u32);
+
+    Ok(Duration::seconds(avg_diff as i64))
+}
+
+pub fn time_to_mine(block_height: u64) -> Result<Duration> {
+    Ok(get_block_time(block_height)? - get_block_time(block_height - 1)?)
+}
+
+/**
+ * Attempts to use average time to mine a block to guess when the next block will be mined
+ */
+pub fn guess_time_to_mine_next_block() -> Result<Duration> {
+    let rpc = &*RPC_CLIENT;
+    let tip = rpc.get_block_count()?;
+    let avg_time = avg_time_to_mine(tip)?;
 
-// TODO: Task 1
-fn time_to_mine(block_height: u64) -> Duration {
-    // * is a deref operator which invokes the Deref trait of the type RPC_CLIENT which was created
-    // when the lazy macro is expanded
-    // if a value has a static lifetime then it means that value lives as long as the program lives
-    let rpc_client: &Client = &*RPC_CLIENT;
-    rpc_client.get_block_hash(234);
-    todo!()
+    let now = Utc::now().timestamp();
+
+    let time_to_mine =
+        avg_time - Duration::seconds(now - get_block_time(tip)?.num_seconds() as i64);
+    Ok(time_to_mine)
+}
+
+pub fn number_of_transactions(block_height: u64) -> Result<u16> {
+    let block = get_block_by_height(block_height)?;
+    Ok(block.txdata.len() as u16)
 }
 
-// TODO: Task 2
-fn number_of_transactions(block_height: u64) -> u16 {
-    let some_value = Box::new(4 as u32);
-    todo!()
+pub fn get_chain() -> Result<Network> {
+    let rpc = &*RPC_CLIENT;
+    let chain = rpc.get_blockchain_info()?;
+    Ok(chain.chain)
+}
+
+use clap::{Parser, Subcommand};
+
+#[derive(Parser)]
+#[command(name = "bitcoin-rpc-cli")]
+#[command(version = "0.1.0")]
+#[command(about = "A CLI for the Bitcoin RPC API", long_about = None)]
+struct Cli {
+    #[command(subcommand)]
+    command: Option<Commands>,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    #[command(about = "Get the current chain")]
+    Chain,
+    #[command(about = "Get the it took to mine a block")]
+    TimeToMine {
+        #[arg(required = true, help = "(numeric, required) The height index")]
+        block_height: u64,
+    },
+    #[command(about = "Get the number of transactions in a block")]
+    NumberOfTransactions {
+        #[arg(required = true, help = "(numeric, required) The height index")]
+        block_height: u64,
+    },
+    #[command(about = "Guess how long until next block is mined")]
+    NextBlock,
+}
+
+// QUESTION: is this the best way to make error handling happen in a single place?
+// if command returns an error then return it to whatever called `call_command`
+fn call_command(command: Commands) -> std::result::Result<(), Box<dyn Error>> {
```

Comment:
Does the error need to be boxed or will it always be a `BitcoinRpcError`?

You can have the main function return a `Result` type if you want. It will just print the debug output of the error message if there's an error.